File: src/main/java/org/tobi29/scapes/engine/opengl/VAOUtility.java
/*
 * Copyright 2012-2015 Tobi29
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.tobi29.scapes.engine.opengl;

import java.util.ArrayList;
import java.util.List;

public class VAOUtility {
    public static VAO createVI(float[] vertex, int[] index,
            RenderType renderType) {
        List<VAO.VAOAttribute> vaoAttributes = new ArrayList<>();
        vaoAttributes
                .add(new VAO.VAOAttribute(OpenGL.VERTEX_ATTRIBUTE, 3, vertex,
                        false, VertexType.HALF_FLOAT));
        return new VAO(vaoAttributes, vertex.length / 3, index, renderType);
    }

    public static VAO createVNI(float[] vertex, float[] normal, int[] index,
            RenderType renderType) {
        List<VAO.VAOAttribute> vaoAttributes = new ArrayList<>();
        vaoAttributes
                .add(new VAO.VAOAttribute(OpenGL.VERTEX_ATTRIBUTE, 3, vertex,
                        false, VertexType.HALF_FLOAT));
        vaoAttributes
                .add(new VAO.VAOAttribute(OpenGL.NORMAL_ATTRIBUTE, 3, normal,
                        true, VertexType.BYTE));
        return new VAO(vaoAttributes, vertex.length / 3, index, renderType);
    }

    public static VAO createVCI(float[] vertex, float[] color, int[] index,
            RenderType renderType) {
        List<VAO.VAOAttribute> vaoAttributes = new ArrayList<>();
        vaoAttributes
                .add(new VAO.VAOAttribute(OpenGL.VERTEX_ATTRIBUTE, 3, vertex,
                        false, VertexType.HALF_FLOAT));
        vaoAttributes.add(new VAO.VAOAttribute(OpenGL.COLOR_ATTRIBUTE, 4, color,
                false, VertexType.UNSIGNED_BYTE));
        return new VAO(vaoAttributes, vertex.length / 3, index, renderType);
    }

    public static VAO createVTI(float[] vertex, float[] texture, int[] index,
            RenderType renderType) {
        List<VAO.VAOAttribute> vaoAttributes = new ArrayList<>();
        vaoAttributes
                .add(new VAO.VAOAttribute(OpenGL.VERTEX_ATTRIBUTE, 3, vertex,
                        false, VertexType.HALF_FLOAT));
        vaoAttributes
                .add(new VAO.VAOAttribute(OpenGL.TEXTURE_ATTRIBUTE, 2, texture,
                        false, VertexType.FLOAT));
        return new VAO(vaoAttributes, vertex.length / 3, index, renderType);
    }

    public static VAO createVTNI(float[] vertex, float[] texture,
            float[] normal, int[] index, RenderType renderType) {
        List<VAO.VAOAttribute> vaoAttributes = new ArrayList<>();
        vaoAttributes
                .add(new VAO.VAOAttribute(OpenGL.VERTEX_ATTRIBUTE, 3, vertex,
                        false, VertexType.HALF_FLOAT));
        vaoAttributes
                .add(new VAO.VAOAttribute(OpenGL.TEXTURE_ATTRIBUTE, 2, texture,
                        false, VertexType.FLOAT));
        vaoAttributes
                .add(new VAO.VAOAttribute(OpenGL.NORMAL_ATTRIBUTE, 3, normal,
                        true, VertexType.BYTE));
        return new VAO(vaoAttributes, vertex.length / 3, index, renderType);
    }

    public static VAO createVCTI(float[] vertex, float[] color, float[] texture,
            int[] index, RenderType renderType) {
        List<VAO.VAOAttribute> vaoAttributes = new ArrayList<>();
        vaoAttributes
                .add(new VAO.VAOAttribute(OpenGL.VERTEX_ATTRIBUTE, 3, vertex,
                        false, VertexType.HALF_FLOAT));
        vaoAttributes.add(new VAO.VAOAttribute(OpenGL.COLOR_ATTRIBUTE, 4, color,
                true, VertexType.UNSIGNED_BYTE));
        vaoAttributes
                .add(new VAO.VAOAttribute(OpenGL.TEXTURE_ATTRIBUTE, 2, texture,
                        false, VertexType.FLOAT));
        return new VAO(vaoAttributes, vertex.length / 3, index, renderType);
    }

    public static VAO createVCTNI(float[] vertex, float[] color,
            float[] texture, float[] normal, int[] index,
            RenderType renderType) {
        List<VAO.VAOAttribute> vaoAttributes = new ArrayList<>();
        vaoAttributes
                .add(new VAO.VAOAttribute(OpenGL.VERTEX_ATTRIBUTE, 3, vertex,
                        false, VertexType.HALF_FLOAT));
        vaoAttributes.add(new VAO.VAOAttribute(OpenGL.COLOR_ATTRIBUTE, 4, color,
                true, VertexType.UNSIGNED_BYTE));
        vaoAttributes
                .add(new VAO.VAOAttribute(OpenGL.TEXTURE_ATTRIBUTE, 2, texture,
                        false, VertexType.FLOAT));
        vaoAttributes
                .add(new VAO.VAOAttribute(OpenGL.NORMAL_ATTRIBUTE, 3, normal,
                        true, VertexType.BYTE));
        return new VAO(vaoAttributes, vertex.length / 3, index, renderType);
    }
}

File: Engine/src/main/kotlin/org/tobi29/scapes/engine/graphics/VAOUtility.kt
/*
 * Copyright 2012-2017 Tobi29
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.tobi29.scapes.engine.graphics

fun GraphicsObjectSupplier.createV(vertex: FloatArray,
                                   renderType: RenderType): Model {
    val attributes = ArrayList<ModelAttribute>()
    attributes.add(ModelAttribute(GL.VERTEX_ATTRIBUTE, 3, vertex, false,
            0, VertexType.HALF_FLOAT))
    return createModelFast(attributes, vertex.size / 3, renderType)
}

fun GraphicsObjectSupplier.createVI(vertex: FloatArray,
                                    index: IntArray,
                                    renderType: RenderType): Model {
    val attributes = ArrayList<ModelAttribute>()
    attributes.add(ModelAttribute(GL.VERTEX_ATTRIBUTE, 3, vertex, false,
            0, VertexType.HALF_FLOAT))
    return createModelStatic(attributes, vertex.size / 3, index, renderType)
}

fun GraphicsObjectSupplier.createVN(vertex: FloatArray,
                                    normal: FloatArray,
                                    renderType: RenderType): Model {
    val attributes = ArrayList<ModelAttribute>()
    attributes.add(ModelAttribute(GL.VERTEX_ATTRIBUTE, 3, vertex, false,
            0, VertexType.HALF_FLOAT))
    attributes.add(ModelAttribute(GL.NORMAL_ATTRIBUTE, 3, normal, true, 0,
            VertexType.BYTE))
    return createModelFast(attributes, vertex.size / 3, renderType)
}

fun GraphicsObjectSupplier.createVNI(vertex: FloatArray,
                                     normal: FloatArray,
                                     index: IntArray,
                                     renderType: RenderType): Model {
    val attributes = ArrayList<ModelAttribute>()
    attributes.add(ModelAttribute(GL.VERTEX_ATTRIBUTE, 3, vertex, false,
            0, VertexType.HALF_FLOAT))
    attributes.add(ModelAttribute(GL.NORMAL_ATTRIBUTE, 3, normal, true, 0,
            VertexType.BYTE))
    return createModelStatic(attributes, vertex.size / 3, index, renderType)
}

fun GraphicsObjectSupplier.createVC(vertex: FloatArray,
                                    color: FloatArray,
                                    renderType: RenderType): Model {
    val attributes = ArrayList<ModelAttribute>()
    attributes.add(ModelAttribute(GL.VERTEX_ATTRIBUTE, 3, vertex, false,
            0, VertexType.HALF_FLOAT))
    attributes.add(ModelAttribute(GL.COLOR_ATTRIBUTE, 4, color, true, 0,
            VertexType.UNSIGNED_BYTE))
    return createModelFast(attributes, vertex.size / 3, renderType)
}

fun GraphicsObjectSupplier.createVCI(vertex: FloatArray,
                                     color: FloatArray,
                                     index: IntArray,
                                     renderType: RenderType): Model {
    val attributes = ArrayList<ModelAttribute>()
    attributes.add(ModelAttribute(GL.VERTEX_ATTRIBUTE, 3, vertex, false,
            0, VertexType.HALF_FLOAT))
    attributes.add(ModelAttribute(GL.COLOR_ATTRIBUTE, 4, color, true, 0,
            VertexType.UNSIGNED_BYTE))
    return createModelStatic(attributes, vertex.size / 3, index, renderType)
}

fun GraphicsObjectSupplier.createVT(vertex: FloatArray,
                                    texture: FloatArray,
                                    renderType: RenderType): Model {
    val attributes = ArrayList<ModelAttribute>()
    attributes.add(ModelAttribute(GL.VERTEX_ATTRIBUTE, 3, vertex, false,
            0, VertexType.HALF_FLOAT))
    attributes.add(ModelAttribute(GL.TEXTURE_ATTRIBUTE, 2, texture, false,
            0, VertexType.FLOAT))
    return createModelFast(attributes, vertex.size / 3, renderType)
}

fun GraphicsObjectSupplier.createVTI(vertex: FloatArray,
                                     texture: FloatArray,
                                     index: IntArray,
                                     renderType: RenderType): Model {
    val attributes = ArrayList<ModelAttribute>()
    attributes.add(ModelAttribute(GL.VERTEX_ATTRIBUTE, 3, vertex, false,
            0, VertexType.HALF_FLOAT))
    attributes.add(ModelAttribute(GL.TEXTURE_ATTRIBUTE, 2, texture, false,
            0, VertexType.FLOAT))
    return createModelStatic(attributes, vertex.size / 3, index, renderType)
}

fun GraphicsObjectSupplier.createVTN(vertex: FloatArray,
                                     texture: FloatArray,
                                     normal: FloatArray,
                                     renderType: RenderType): Model {
    val attributes = ArrayList<ModelAttribute>()
    attributes.add(ModelAttribute(GL.VERTEX_ATTRIBUTE, 3, vertex, false,
            0, VertexType.HALF_FLOAT))
    attributes.add(ModelAttribute(GL.TEXTURE_ATTRIBUTE, 2, texture, false,
            0, VertexType.FLOAT))
    attributes.add(ModelAttribute(GL.NORMAL_ATTRIBUTE, 3, normal, true, 0,
            VertexType.BYTE))
    return createModelFast(attributes, vertex.size / 3, renderType)
}

fun GraphicsObjectSupplier.createVTNI(vertex: FloatArray,
                                      texture: FloatArray,
                                      normal: FloatArray,
                                      index: IntArray,
                                      renderType: RenderType): Model {
    val attributes = ArrayList<ModelAttribute>()
    attributes.add(ModelAttribute(GL.VERTEX_ATTRIBUTE, 3, vertex, false,
            0, VertexType.HALF_FLOAT))
    attributes.add(ModelAttribute(GL.TEXTURE_ATTRIBUTE, 2, texture, false,
            0, VertexType.FLOAT))
    attributes.add(ModelAttribute(GL.NORMAL_ATTRIBUTE, 3, normal, true, 0,
            VertexType.BYTE))
    return createModelStatic(attributes, vertex.size / 3, index, renderType)
}

fun GraphicsObjectSupplier.createVCT(vertex: FloatArray,
                                     color: FloatArray,
                                     texture: FloatArray,
                                     renderType: RenderType): Model {
    val attributes = ArrayList<ModelAttribute>()
    attributes.add(ModelAttribute(GL.VERTEX_ATTRIBUTE, 3, vertex, false,
            0, VertexType.HALF_FLOAT))
    attributes.add(ModelAttribute(GL.COLOR_ATTRIBUTE, 4, color, true, 0,
            VertexType.UNSIGNED_BYTE))
    attributes.add(ModelAttribute(GL.TEXTURE_ATTRIBUTE, 2, texture, false,
            0, VertexType.FLOAT))
    return createModelFast(attributes, vertex.size / 3, renderType)
}

fun GraphicsObjectSupplier.createVCTI(vertex: FloatArray,
                                      color: FloatArray,
                                      texture: FloatArray,
                                      index: IntArray,
                                      renderType: RenderType): Model {
    val attributes = ArrayList<ModelAttribute>()
    attributes.add(ModelAttribute(GL.VERTEX_ATTRIBUTE, 3, vertex, false,
            0, VertexType.HALF_FLOAT))
    attributes.add(ModelAttribute(GL.COLOR_ATTRIBUTE, 4, color, true, 0,
            VertexType.UNSIGNED_BYTE))
    attributes.add(ModelAttribute(GL.TEXTURE_ATTRIBUTE, 2, texture, false,
            0, VertexType.FLOAT))
    return createModelStatic(attributes, vertex.size / 3, index, renderType)
}

fun GraphicsObjectSupplier.createVCTN(vertex: FloatArray,
                                      color: FloatArray,
                                      texture: FloatArray,
                                      normal: FloatArray,
                                      renderType: RenderType): Model {
    val attributes = ArrayList<ModelAttribute>()
    attributes.add(ModelAttribute(GL.VERTEX_ATTRIBUTE, 3, vertex, false,
            0, VertexType.HALF_FLOAT))
    attributes.add(ModelAttribute(GL.COLOR_ATTRIBUTE, 4, color, true, 0,
            VertexType.UNSIGNED_BYTE))
    attributes.add(ModelAttribute(GL.TEXTURE_ATTRIBUTE, 2, texture, false,
            0, VertexType.FLOAT))
    attributes.add(ModelAttribute(GL.NORMAL_ATTRIBUTE, 3, normal, true, 0,
            VertexType.BYTE))
    return createModelFast(attributes, vertex.size / 3, renderType)
}

fun GraphicsObjectSupplier.createVCTNI(vertex: FloatArray,
                                       color: FloatArray,
                                       texture: FloatArray,
                                       normal: FloatArray,
                                       index: IntArray,
                                       renderType: RenderType): Model {
    val attributes = ArrayList<ModelAttribute>()
    attributes.add(ModelAttribute(GL.VERTEX_ATTRIBUTE, 3, vertex, false,
            0, VertexType.HALF_FLOAT))
    attributes.add(ModelAttribute(GL.COLOR_ATTRIBUTE, 4, color, true, 0,
            VertexType.UNSIGNED_BYTE))
    attributes.add(ModelAttribute(GL.TEXTURE_ATTRIBUTE, 2, texture, false,
            0, VertexType.FLOAT))
    attributes.add(ModelAttribute(GL.NORMAL_ATTRIBUTE, 3, normal, true, 0,
            VertexType.BYTE))
    return createModelStatic(attributes, vertex.size / 3, index, renderType)
}
